### PR TITLE
Ratelimit invites by room and target user

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/config/ratelimiting.py [hunk 1]
```diff
@@ -107,6 +107,15 @@ def read_config(self, config, **kwargs):
...
         )
 
+        self.rc_invites_per_room = RateLimitConfig(
+            config.get("rc_invites", {}).get("per_room", {}),
+            defaults={"per_second": 0.003, "burst_count": 10},
+        )
+        self.rc_invites_per_user = RateLimitConfig(
+            config.get("rc_invites", {}).get("per_user", {}),
+            defaults={"per_second": 0.003, "burst_count": 5},
+        )
```
Comment (anoadragon453) [inline synapse/config/ratelimiting.py:117]: Should these be `0.03` instead of `0.003`?

File: synapse/handlers/room_member.py [hunk 2]
```diff
@@ -144,6 +155,12 @@ async def _user_left_room(self, target: UserID, room_id: str) -> None:
         """
         raise NotImplementedError()
 
+    def ratelimit_invite(self, room_id: str, invitee_user_id: str):
+        """Ratelimit invites py room and by target user.
```
Comment (anoadragon453) [inline synapse/handlers/room_member.py:159]:
```suggestion
        """Ratelimit invites by room and by target user.
```

File: tests/handlers/test_federation.py [hunk 3]
```diff
@@ -191,6 +191,97 @@ def test_rejected_state_event_state(self):
...
+                    room_version,
+                )
+            )
+
+        self.get_failure(
+            self.handler.on_invite_request(
+                other_server, create_invite_for("@user-4:test"), room_version,
+            ),
+            exc=LimitExceededError,
+        )
```
Comment (anoadragon453) [inline tests/handlers/test_federation.py:239]: So my local server would send a 429 back over federation to the remote server? I'm guessing this would just result in a failure on the remote side (possibly without a useful message)?

File: changelog.d/9258.feature [hunk 4]
```diff
@@ -0,0 +1 @@
+Add ratelimites to invites in rooms and to specific users.
```
Comment (anoadragon453) [inline changelog.d/9258.feature:1]:
```suggestion
Add ratelimits to invites in rooms and to specific users.
```

File: synapse/handlers/room_member.py [hunk 5]
```diff
@@ -387,8 +404,11 @@ async def update_membership_locked(
                 raise SynapseError(403, "This room has been blocked on this server")
 
         if effective_membership_state == Membership.INVITE:
+            target_id = target.to_string()
+            self.ratelimit_invite(room_id, target_id)
```
Comment (anoadragon453) [inline synapse/handlers/room_member.py:408]: I wonder whether we should condition this on the `ratelimit` argument to this function. Unfortunately doing so would allow users to mass invite by supplying a long list of users to `/createRoom`. However not doing so might cause the server to get rate-limited accidentally?

File: synapse/handlers/room_member.py (hunk 2 above)
Comment (dkasak) [inline synapse/handlers/room_member.py:159]:
```suggestion
        """Ratelimit invites by room and by target user.
```

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
A few edge cases to think about, but on the whole looks good.

File: tests/handlers/test_federation.py (hunk 3 above)
Comment (erikjohnston) [reply tests/handlers/test_federation.py:239]: I believe the 429 will be proxied through to the client.

File: synapse/handlers/room_member.py (hunk 5 above)
Comment (erikjohnston) [reply synapse/handlers/room_member.py:408]: Oh good point. I can also limit the number of invites in `/createRoom` by the burst count.

File: synapse/config/ratelimiting.py [hunk 6]
```diff
@@ -107,6 +107,15 @@ def read_config(self, config, **kwargs):
             defaults={"per_second": 0.003, "burst_count": 5},
         )
 
+        self.rc_invites_per_room = RateLimitConfig(
+            config.get("rc_invites", {}).get("per_room", {}),
+            defaults={"per_second": 0.3, "burst_count": 10},
+        )
+        self.rc_invites_per_user = RateLimitConfig(
+            config.get("rc_invites", {}).get("per_user", {}),
+            defaults={"per_second": 0.003, "burst_count": 5},
```
Comment (anoadragon453) [inline synapse/config/ratelimiting.py:116]: This doesn't match the sample config.

File: synapse/config/ratelimiting.py [hunk 7]
```diff
@@ -174,6 +185,14 @@ def generate_config_section(self, **kwargs):
         #rc_3pid_validation:
         #  per_second: 0.003
         #  burst_count: 5
+        #
+        #rc_invites:
+        #  per_room:
+        #    per_second: 0.3
+        #    burst_count: 10
+        #  per_user:
+        #    per_second: 0.03
```
Comment (anoadragon453) [inline synapse/config/ratelimiting.py:194]: Did you mean for these two `per_second`s to be different?

File: tests/handlers/test_federation.py (hunk 3 above)
Comment (anoadragon453) [reply tests/handlers/test_federation.py:239]: Looks like a `HttpResponseException` with code 429 will get raised:

https://github.com/matrix-org/synapse/blob/2649d545a551dd126d73d34a6e3172916ea483e0/synapse/federation/federation_client.py#L783-L805

So it won't be a `LimitExceededError` and won't have `retry_after_ms` nor `M_LIMIT_EXCEEDED` but it miiiight be handled correctly by clients?

We may just want to catch `429`, craft and raise a `LimitExceededError` to be extra safe.

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Getting there!

File: synapse/config/ratelimiting.py (hunk 6 above)
Comment (erikjohnston) [reply synapse/config/ratelimiting.py:116]: SIGH I THOUGHT I DOUBLE CHECKED

File: synapse/config/ratelimiting.py (hunk 7 above)
Comment (erikjohnston) [reply synapse/config/ratelimiting.py:194]: Yes, so we want a relatively high `per_room` ratelimit to allow room upgrades, but we can limit the number of invites a user receives a lot (one every 5 minutes for `0.003`)

File: tests/handlers/test_federation.py (hunk 3 above)
Comment (erikjohnston) [reply tests/handlers/test_federation.py:239]: Oh thank you. Have fixed this now

File: synapse/federation/federation_client.py [hunk 8]
```diff
@@ -810,7 +810,7 @@ async def _do_send_invite(
                         "User's homeserver does not support this room version",
                         Codes.UNSUPPORTED_ROOM_VERSION,
                     )
-            elif e.code == 403:
+            elif e.code in (403, 429):
                 raise e.to_synapse_error()
```
Comment (anoadragon453) [inline synapse/federation/federation_client.py:814]: Cool, thanks for added. This won't pass through `retry_after_ms` but I presume the client will have some default amount of time they'll wait for.

**Review by anoadragon453 — APPROVED** [review]
lgtm!

File: synapse/federation/federation_client.py (hunk 8 above)
Comment (erikjohnston) [reply synapse/federation/federation_client.py:814]: Oh, hmm, why on earth don't we proxy through all the fields??

I don't think clients actually care about `retry_after_ms`...

File: docs/sample_config.yaml [hunk 9]
```diff
@@ -862,6 +864,14 @@ log_config: "CONFDIR/SERVERNAME.log.config"
...
 #  per_second: 0.003
 #  burst_count: 5
+#
+#rc_invites:
+#  per_room:
+#    per_second: 0.3
+#    burst_count: 10
+#  per_user:
+#    per_second: 0.003
+#    burst_count: 5
```
Comment (richvdh) [inline docs/sample_config.yaml:874]: if this is a ratelimit per recipient, does that mean that a spammer can DoS a given recipient so that they can't receive any invites from genuine users any more? what happens to genuine invites - are they just lost somehow?

Comment (richvdh) [reply docs/sample_config.yaml:874]: (could this sort of info be added to the comments in the config file, please?)

Comment (erikjohnston) [reply docs/sample_config.yaml:874]: Hmm, true. This does mean In some ways this is somewhat by design, the idea is that you don't want to have users inundated by invites from a lot of different fake accounts. The invites will see the 429 though, so at least its a visible thing. I guess we could ratelimit by `(sending_server, target_user)` for inbound remote invites so that a spammer would have to have accounts on different servers?

Comment (richvdh) [reply docs/sample_config.yaml:874]: maybe. Honestly, I'm kinda questioning how useful this limit is as it stands.
